esp32/mods/esp_espnow.c

```diff
@@ -0,0 +1,359 @@
+    if (re_add) {
+        // workaround for calling esp_now_mod_peer() to 
+        // change encryption status will crash the system
+        peer.encrypt = encrypt;
+        ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+        ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    } else {
+        ESPNOW_EXCEPTIONS(esp_now_mod_peer(&peer));
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(espnow_lmk_obj, espnow_lmk);
+
+// add_peer(peer_mac, [local_key])
+STATIC mp_obj_t espnow_add_peer(size_t n_args, const mp_obj_t *args) {
+    esp_now_peer_info_t peer = {0};
+    _get_bytes(args[1], ESP_NOW_ETH_ALEN, peer.peer_addr);
+    if (n_args > 2) {
+        _get_bytes(args[2], ESP_NOW_KEY_LEN, peer.lmk);
+        peer.encrypt = 1;
+    }
+
+    ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_add_peer_obj, 2, 3, espnow_add_peer);
+
+// del_peer(peer_mac)
+STATIC mp_obj_t espnow_del_peer(mp_obj_t self_in, mp_obj_t addr) {
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf = (const uint8_t *)mp_obj_str_get_data(addr, &addr_len);
+    if (addr_len != ESP_NOW_ETH_ALEN) mp_raise_ValueError("addr invalid");
+    ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(espnow_del_peer_obj, espnow_del_peer);
+
+// this workaround enables ESP32 to send from whatever IF that is
+// active
+// if_id == wifi_mode - 1
+#define IS_IF_AVAILABLE(mode, if_id) ({ (mode & (if_id+1)) != 0; })
+#define AVAILABLE_IF(mode) ({                                       \
+    int if_id = -1;                                                 \
+    for (int i=WIFI_MODE_STA; i<=WIFI_MODE_AP; i++) {               \
+        if (mode & i) {                                             \
+            if_id = i-1;                                            \
+        }                                                           \
+    }                                                               \
+    if_id;                                                          \
+}) 
+
+// this hack is used to directly reference the flag from modwlan.c Line 123
+// The variable has been changed to not static to make it visble outside the wlan module.
+//  TODO - change to public call
+
+//extern bool wifi_started;
+
+STATIC mp_obj_t espnow_send(mp_obj_t self_in, mp_obj_t addr, mp_obj_t msg) {
+//    if (!wifi_started) goto espnow_wifi_err;
+
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf;
+    mp_uint_t msg_len;
+    const uint8_t *msg_buf = (const uint8_t *)mp_obj_str_get_data(msg, &msg_len);
+    if (msg_len > ESP_NOW_MAX_DATA_LEN) mp_raise_ValueError("msg too long");
+
+    wifi_mode_t mode;
+    ESPNOW_EXCEPTIONS(esp_wifi_get_mode(&mode));
+    bool first = true;
+    int new_if = -1;
+    if (addr == mp_const_none) {
+        // send to all
+        esp_now_peer_info_t peer;
+        esp_err_t e = esp_now_fetch_peer(true, &peer);
```

Comment:
In this case, wouldn't it be enough to call esp_now_send() with NULL as peer_addr? As far I understand in that case it sends out the packet to all available peers.